src/UserSpaceInstrumentation/OrbitUserSpaceInstrumentation.cpp

```diff
@@ -73,47 +56,80 @@ class LockFreeUserSpaceInstrumentationEventProducer
 
  protected:
   [[nodiscard]] orbit_grpc_protos::ProducerCaptureEvent* TranslateIntermediateEvent(
-      FunctionCallEvent&& raw_event, google::protobuf::Arena* arena) override {
+      FunctionEntryExitVariant&& raw_event, google::protobuf::Arena* arena) override {
     auto* capture_event =
         google::protobuf::Arena::CreateMessage<orbit_grpc_protos::ProducerCaptureEvent>(arena);
-    auto* function_call = capture_event->mutable_function_call();
-    function_call->set_pid(raw_event.pid);
-    function_call->set_tid(raw_event.tid);
-    function_call->set_function_id(raw_event.function_id);
-    function_call->set_duration_ns(raw_event.duration_ns);
-    function_call->set_end_timestamp_ns(raw_event.end_timestamp_ns);
+
+    std::visit(
+        [capture_event](auto&& raw_event) {
+          using DecayedEventType = std::decay_t<decltype(raw_event)>;
+          if constexpr (std::is_same_v<DecayedEventType, orbit_grpc_protos::FunctionEntry>) {
+            orbit_grpc_protos::FunctionEntry* function_entry =
+                capture_event->mutable_function_entry();
+            *function_entry = std::forward<decltype(raw_event)>(raw_event);
+          } else if constexpr (std::is_same_v<DecayedEventType, orbit_grpc_protos::FunctionExit>) {
+            orbit_grpc_protos::FunctionExit* function_exit = capture_event->mutable_function_exit();
+            *function_exit = std::forward<decltype(raw_event)>(raw_event);
+          } else {
+            static_assert(always_false_v<DecayedEventType>, "Non-exhaustive visitor");
+          }
+        },
+        std::move(raw_event));
+
     return capture_event;
   }
+
+ private:
+  template <class>
+  [[maybe_unused]] static constexpr bool always_false_v = false;
```

Comment:
Is it possible to make this a local in the lambda?

Reply:
Unfortunately no because of the template, I already tried actually.